dbms/src/Storages/KVStore/MultiRaft/Disagg/FastAddPeerContext.cpp

```diff
@@ -110,7 +110,7 @@ std::pair<UInt64, ParsedCheckpointDataHolderPtr> FastAddPeerContext::getNewerChe
     return std::make_pair(cache_seq, checkpoint_data);
 }
 
-CheckpointIngestInfoPtr FastAddPeerContext::getOrRestoreCheckpointIngestInfo(
+std::optional<CheckpointIngestInfoPtr> FastAddPeerContext::getOrRestoreCheckpointIngestInfo(
```

Comment:
Will it cost too much if I use `std::optional<CheckpointIngestInfoPtr>`? Because for safety issue, I prefer this one, since dereferencing a nullptr result in a panic which is hard to diagnose.

Reply:
> there is no difference with dereferenceing a `nullptr`

It actually does with common usage. Given the below example, if we call `.value` on `std::nullopt`, it will **throw** an `std::bad_optional_access` which [is not UB](https://en.cppreference.com/w/cpp/utility/optional/value). However, it we call `operator->` on `nullptr`, it will panic. In TiFlash, the `BaseDaemon` handles exceptions better than segment faults.
```
#include <optional>
#include <cstring>
#include <cstdio>

void f() {
    try {
        std::optional<int> x = std::nullopt;
        printf("%d\n", x.value());
    }
    catch (std::bad_optional_access & e) {
        printf("it throws in f\n");
    }
}

void g() {
    try {
        int * x = nullptr;
        printf("%d\n", *x);
    } catch (...){
        printf("it throws in g\n");
    }
}

int main() {
    f();
    g();
}
```

> I think `shared_ptr`/`unique_ptr` and `optional` are redundant

I think it depends. In rust, it is very common to use `Option<NotNull<T>>` to [ensure safety](https://stackoverflow.com/questions/54195517/should-we-use-option-or-ptrnull-to-represent-a-null-pointer-in-rust). In C++, there is no such thing like `NotNull`, however, I think it is also proper to use `std::option<std::shared_ptr<T>>` in API methods which explicitly claim the pointer could be null

Reply:
OK, I agree with you now. So how about we introduce `not_null` and some optional wrapper that totally prevents UB behaviors for these cases?

Reply:
> In TiFlash, the BaseDaemon handles exceptions better than segment faults.

I don't agree with this. The BaseDaemon only handles `DB::Exception` better than segment faults. But it handles the exceptions from C++ std lib badly. We can get the stack backtrace when `DB::Exception` is thrown or a null pointer is being accessed, but no stack backtrace when std exceptions are thrown.

Reply:
it is changed back to std::shared_ptr